docs/core/hd-keys.md

```diff
@@ -0,0 +1,131 @@
+---
+id: hd-keys
+title: HD key generation
+---
+
+In modern Bitcoin wallets, users only need to write down
+a sequence of words, and that sequence is a complete backup
+of their wallet. This is thanks to what's called Hierarchical
+Deterministic key generation. In short, every wallet using HD
+key generation has a root seed for each wallet, and this
+seed can be used to generate an infinite amount of later
+private and public keys. This is done in a standardized manner,
+so different wallets can operate with the same standard.
+
+> If you want to jump into the details of how this work,
+> you should check out
+> [BIP 32](https://github.com/bitcoin/bips/blob/master/bip-0032.mediawiki).
+
+Bitcoin-S supports generating keys in this fashion. Here's a
+full example of how to obtain a wallet seed, and then
+use that to generate further private and public keys:
+
+Here's an example:
+
+```scala mdoc:to-string
+import scodec.bits._
+import org.bitcoins.core.crypto._
+import org.bitcoins.core.hd._
+
+// the length of the entropy bit vector determine
+// how long our phrase ends up being
+// 256 bits of entropy results in 24 words
+val entropy: BitVector = MnemonicCode.getEntropy256Bits
+
+val mnemonicCode = MnemonicCode.fromEntropy(entropy)
+
+mnemonicCode.words // the phrase the user should write down
+
+// the password argument is an optional, extra security
+// measure. all MnemonicCode instances will give you a
+// valid BIP39 seed, but different passwords will give
+// you different seeds. So you could have as many wallets
+// from the same seed as you'd like, by simply giving them
+// different passwords.
+val bip39Seed = BIP39Seed.fromMnemonic(mnemonicCode,
+                                       password = "secret password")
+
+val xpriv = ExtPrivateKey.fromBIP39Seed(ExtKeyVersion.SegWitMainNetPriv,
+                                        bip39Seed)
+val xpub = xpriv.extPublicKey
+
+// you can now use the generated xpriv to derive further
+// private or public keys
+
+// this can be done with BIP89 paths (called SegWitHDPath in bitcoin-s)
+val segwitPath = SegWitHDPath.fromString("m/84'/0'/0'/0/0")
+
+// alternatively:
+val otherSegwitPath =
+SegWitHDPath(HDCoinType.Bitcoin,
+             accountIndex = 0,
+             HDChainType.External,
+             addressIndex = 0)
+
+segwitPath == otherSegwitPath
```

Comment:
either wrap in `assert` or comment `// true` for clarity

Reply:
So this is one of the cool things about mdoc (https://github.com/scalameta/mdoc), it typechecks and even runs the code before publishing. So `// true` gets added at the bottom by the compiler

Reply:
gotchya, very cool!